### PR TITLE
Adding XAML Extensions

### LINKED ISSUE #472: Add XAML Extensions For Standard Controls
## 🏗 Enhancement Proposal

Rather than having to generate a ton of Commands inside your ViewModel, XAML extensions should be added that wrap the common player tasks.

### Pitch

- Add XmlnsDefinition to make it a little nicer to bring in the Media Manager resources

```cs
[assembly: XmlnsPrefix("http://baseflow.com/mediamanager", "media")]
[assembly: XmlnsDefinition("http://baseflow.com/mediamanager", "MediaManager.Forms")]
[assembly: XmlnsDefinition("http://baseflow.com/mediamanager", "MediaManager.Forms.Xaml")]
```

- Add XAML Extensions to provide a Command to execute free of a ViewModel
  - PauseExtension
  - PlayExtension
  - PlayNextExtension
  - PlayPauseExtension
  - PlayPreviousExtension
  - PlayPreviousOrSeekToStartExtension
  - SeekToStartExtension
  - StepBackwardExtension
  - StepForwardExtension
  - StopExtension
  - ToggleRepeatExtension
  - ToggleShuffleExtension

```xml
<ContentPage xmlns:media="http://baseflow.com/mediamanager">
  <Button Text="<<"
          Command="{media:PlayPrevious}" />
  <Button Text="> /  ||"
          Command="{media:PlayPause}" />
  <Button Text=">>"
          Command="{media:PlayNext}" />
```

### Platforms affected (mark all that apply)
- [ ] :iphone: iOS
- [ ] :robot: Android
- [ ] :checkered_flag: WPF
- [ ] :earth_americas: UWP
- [ ] :apple: MacOS
- [ ] :tv: tvOS
- [x] :monkey: Xamarin.Forms

### PR DESCRIPTION
**Pull request by dansiegel**
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)

Adds new XAML Extensions to reduce the number of commands and logic required in your ViewModel

### :arrow_heading_down: What is the current behavior?

Currently all common player controls must be executed from a Command added to the ViewModel or an event handler in Code Behind.

```cs
public class PlayMediaViewModel
{
    public PlayMediaViewModel(IMediaManager mediaManager)
    {
        PlayPauseCommand = new Command(() => mediaManager.PlayPause());
        PlayPreviousCommand = new Command(() => mediaManager.PlayPrevious());
        PlayNextCommand = new Command(() => mediaManager.PlayNext());
    }

    public Command PlayPauseCommand { get; }
    public Command PlayPreviousCommand { get; }
    public Command PlayNextCommand { get; }
}
```

```xml
<Button Text="<<" Command="{Binding PlayPreviousCommand}" />
<Button Text="> / ||" Command="{Binding PlayPauseCommand}" />
<Button Text=">>" Command="{Binding PlayNextCommand}" />
```

### :new: What is the new behavior (if this is a feature change)?

- Adds defined Xmlns Definition `http://baseflow.com/mediamanager`

- Simply reference one of the new XAML extensions for the Command of the Player Control you wish to use.

```xml
<ContentPage xmlns:media="http://baseflow.com/mediamanager">
  <Button Text="<<"
          Command="{media:PlayPrevious}" />
  <Button Text="> /  ||"
          Command="{media:PlayPause}" />
  <Button Text=">>"
          Command="{media:PlayNext}" />
```

### :boom: Does this PR introduce a breaking change?

No

### :bug: Recommendations for testing

n/a

### :memo: Links to relevant issues/docs

fixes #472

### :thinking: Checklist before submitting

- [x] All projects build
- [x] Follows style guide lines 
- [ ] Relevant documentation was updated
- [x] Rebased onto current develop

### CODE REVIEW

File: MediaManager.Forms/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+﻿using Xamarin.Forms;
+
+[assembly: XmlnsPrefix("http://baseflow.com/mediamanager", "media")]
```
Comment (martijn00) [inline MediaManager.Forms/Properties/AssemblyInfo.cs]: MediaManager uses SDK-style project and will generate the assembly info. This makes the build fail i think. Could you change the name into `AssemblyInfoAttributes.cs`? Just like: https://github.com/MvvmCross/MvvmCross/blob/develop/MvvmCross.Forms/AssemblyInfoAttributes.cs

File: MediaManager.Forms/Xaml/StopExtension.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public class StopExtension : MediaExtensionBase
+    {
+        protected override bool CanExecute() => 
+            MediaManager.State == MediaPlayerState.Playing;
+
+        protected override void Execute()
+        {
+            MediaManager.Stop();
+            MediaManager.PlayPreviousOrSeekToStart();
```
Comment (martijn00) [inline MediaManager.Forms/Xaml/StopExtension.cs]: Should this not only contain Stop?

**Comment by martijn00** [comment]
@dansiegel thanks a lot! I really like this approach!

File: MediaManager.Forms/Xaml/StopExtension.cs (hunk 2 above)
Comment (dansiegel) [reply MediaManager.Forms/Xaml/StopExtension.cs]: yes looks like I missed that in the cleanup... thanks for the catch.

File: MediaManager.Forms/Properties/AssemblyInfo.cs (hunk 1 above)
Comment (dansiegel) [reply MediaManager.Forms/Properties/AssemblyInfo.cs]: SDK Style shouldn't have anything to do with that. By convention an SDK style projects generate the file with all of the base Version/Meta-Data attributes as {ProjectName}.AssemblyInfo.cs in the Intermediate Output Folder (i.e. `obj/{Configuration}`)… That said looking at that build output from AppVeyor it seems that they're doing something really weird that is causing a conflict.

Comment (martijn00) [reply MediaManager.Forms/Properties/AssemblyInfo.cs]: I had the same problem with MvvmCross and am pretty sure that if you rename the file it will work.

Comment (dansiegel) [reply MediaManager.Forms/Properties/AssemblyInfo.cs]: No problem.. like I mentioned it looks to me that AppVeyor was not happy with it. I renamed this per your suggestion

**Comment by martijn00** [comment]
Thanks for helping out!

**Pull request merged**